src/module_1/module_1_meteo_api.py

```diff
@@ -1,7 +1,227 @@
+        print(f"Http Error: {errh}")
+    except requests.exceptions.ConnectionError as errc:
+        print(f"Error Connecting: {errc}")
+    except requests.exceptions.Timeout as errt:
+        print(f"Timeout Error: {errt}")
+    except requests.exceptions.RequestException as err:
+        print(f"Error: {err}")
+
+
+def get_data_meteo_api(city, VARIABLES):
+    """
+    Connect to an API using the requests library.
+
+    :param city: URL of the API endpoint.
+    """
+
+    # Define Climate API URL
+    API_URL = "https://climate-api.open-meteo.com/v1/climate?"
+
+    # Define coordinate of available cities
+    COORDINATES = {
+        "Madrid": {"latitude": 40.416775, "longitude": -3.703790},
+        "London": {"latitude": 51.507351, "longitude": -0.127758},
+        "Rio": {"latitude": -22.906847, "longitude": -43.172896},
+    }
+
+    # Define variables to retrieve from API
+    # VARIABLES = ["temperature_2m_mean", "precipitation_sum", "soil_moisture_0_to_10cm_mean"]
+
+    # Define API parameters
+    params = {
+        "latitude": COORDINATES[city]["latitude"],
+        "longitude": COORDINATES[city]["longitude"],
+        "start_date": "1950-01-01",
+        "end_date": "2050-12-31",
+        "models": [
+            "CMCC_CM2_VHR4",
+            "FGOALS_f3_H",
+            "HiRAM_SIT_HR",
+            "MRI_AGCM3_2_S",
+            "EC_Earth3P_HR",
+            "MPI_ESM1_2_XR",
+            "NICAM16_8S",
+        ],
+        "daily": VARIABLES,
+    }
+
+    # Call API
+    response = connect_to_api(API_URL, params=params)
+    data = response.json()
+
+    # Create a DataFrame with the retrieved data
+    columns_df = list(data["daily_units"].keys())
+
+    dicts = {}
+    for column in columns_df:
+        dicts[column] = data["daily"][column]
+
+    df_city_climate = pd.DataFrame(dicts)
+    df_city_climate["city"] = city
+
+    # Save dataframe as csv file
+    data_path = os.getcwd() + "/data"
+    if not os.path.exists(data_path):
+        os.makedirs(data_path)
+
+    save_path = f"{data_path}/{str.lower(city)}_climate.csv"
+    df_city_climate.to_csv(save_path, index=False)
+
+    return df_city_climate
+
+
+def calculate_meteo_mean_max_min_values(df_meteo_0, variables):
+    df_meteo = df_meteo_0.set_index("time")
+
+    for var in variables:
+        columns_var = [col for col in df_meteo.columns if var in col]
+
+        # Get mean, max and mean values of all models available per variable.
+        # Join them together in a single dataframe
+
+        df_join = pd.concat(
+            [
+                pd.DataFrame(
+                    df_meteo[columns_var].mean(axis=1), columns=[var + "_avg"]
+                ),
+                pd.DataFrame(df_meteo[columns_var].max(axis=1), columns=[var + "_max"]),
+                pd.DataFrame(df_meteo[columns_var].min(axis=1), columns=[var + "_min"]),
+            ],
+            axis=1,
+        )
+
+        df_meteo = pd.concat([df_meteo, df_join], axis=1)
```

Comment:
In general it's better to use something else like `merge` if you can avoid `pd.concat`. Also, can `df_meteo` and `df_join` have different number of rows?